File: Components/Strata.Foundation.Core/main/typescript/Utility/Executed.ts
//////////////////////////////////////////////////////////////////////////////
// Executed.ts
//////////////////////////////////////////////////////////////////////////////

import {Optional}              from "./Optional";
import {NullPointerException}  from "./NullPointerException";

export
class Executed
{
    private readonly method:    string;
    private readonly throwable: Error | null;

    private
    constructor(method: string,throwable: Error | null)
    {
        this.method    = method;
        this.throwable = throwable;
    }

    getMethod(): string { return this.method; }

    getException(): Optional<Error> { return Optional.ofNullable<Error>(this.throwable); }

    isSuccess(): boolean { return this.throwable == null; }

    isFailure(): boolean { return this.throwable != null; }

    static of(method: string): Executed;
    static of(method: string,throwable: Error): Executed;
    static of(method: string,throwable?: Error): Executed
    {
        if (method == null)
            throw new NullPointerException("method is null");

        if (throwable !== undefined)
        {
            if (throwable == null)
                return new Executed(method,new NullPointerException("throwable is null"));

            return new Executed(method,throwable);
        }

        return new Executed(method,null);
    }
}

//////////////////////////////////////////////////////////////////////////////

File: Components/Strata.Foundation.Core/main/java/strata/foundation/core/utility/Executed.java
//////////////////////////////////////////////////////////////////////////////
// Executed.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.utility;

import java.util.Objects;
import java.util.Optional;

/**
 * <p>
 * Result produced by an operation that accepts no input and returns
 * no output, capturing the method name and, when the invocation
 * failed, the resulting {@link Throwable}. This covers any
 * fire-and-forget scenario — {@link Runnable} execution, cache
 * invalidation, connection resets, lifecycle hooks, or any void
 * method invoked without arguments.
 * </p><br/>
 * <b>Examples</b><br/>
 * <pre>
 * Executed result = Executed.of("refresh");
 *
 * if (result.isFailure())
 *     log(result.getException());
 * </pre>
 */
public
class Executed
{
    private final String    method;
    private       Throwable throwable;

    public
    Executed(String method)
    {
        this.method    = Objects.requireNonNull(method,"method");
        this.throwable = null;
    }

    public
    Executed(String method,Throwable throwable)
    {
        this.method = Objects.requireNonNull(method,"method");

        try
        {
            this.throwable =
                Objects.requireNonNull(throwable,"throwable");
        }
        catch (Exception e)
        {
            this.throwable = e;
        }
    }

    public String
    getMethod() { return method; }

    public Optional<Throwable>
    getException() { return Optional.ofNullable(throwable); }

    public boolean
    isSuccess() { return this.throwable == null; }

    public boolean
    isFailure() { return this.throwable != null; }

    public static Executed
    of(String method) { return new Executed(method); }

    public static Executed
    of(String method,Throwable throwable) { return new Executed(method,throwable); }
}

//////////////////////////////////////////////////////////////////////////////
